src/h3/api/_api_template.py

```diff
@@ -150,8 +150,18 @@ def uncompact(hexes, res):
 
         return _out_collection(hu)
 
-    def polyfill(geos, res):
-        mv = mvi.polyfill(geos, res)
+    def polyfill_polygon(outer, holes=None, res=9, order='latlng'):
```

Comment:
I'd suggest against defaulting the resolution. This promotes a specific resolution to be the default one, but it's not clear that we can know that for the broader H3 community.

If an API where the user doesn't always need to specify the resolution is what you want, I'd suggest a class where the user provides the resolution once and then it's reused.

Reply:
Yeah, I thought this was a bit awkward too.

This function signature came out of my desire to break up the `holes` from the `outer` LinearRing. I was finding myself constantly making mistakes when the expected input was `[exterior, hole1, hole2, ...]`. For a simple polygon with no holes, it felt awkward to have to write it like `polyfill_polygon([outer])`. 

This all led to making `outer` the first ordered parameter, and adding `holes` as an optional parameter. You can't have non-keyword parameters **after** keyword parameters, so `res` had to have some default value.

I think there are a few different signatures we could use here. I'd love feedback on what you all think would be best.

1. `polyfill_polygon(outer, holes=None, res=9, order='latlng')`
    - current option
2. `polyfill_polygon(outer, holes=None, res=0, order='latlng')`
    - is `res=0` a better default?
3. `polyfill_polygon(outer, holes=None, res=None, order='latlng')`
    - `res` would now be a **required** keyword parameter; we'd raise an error if `res=None`
4. `polyfill_polygon(outer, res, holes=None, order='latlng')`
    - Eh, maybe. is it weird splitting up the `outer` and `holes` parameters?
    - Might be the cleanest option if the most common use case is for Polygons without holes
5. `polyfill_polygon(res, outer, holes=None, order='latlng')`
    - I think this is kind of what you were suggesting @isaacbrodsky. This makes it easy to curry the function, so you could specify `res` just once.
    - It does feel a little unintitive to have `res` first; and it doesn't match our current API.

Reply:
Yes, `polyfill_polygon(res, outer, holes=None, order='latlng')` is closest to my thinking. The other option I'd suggest is `polyfill_polygon(outer, holes, res, order='latlng')`, making `holes` a required argument. I could also see `polyfill_polygon(polygon, res)` where `polygon` is some more complicated type that encapsulates `outer`, `holes`, and `order`.

Reply:
I'm leaning towards option 4, as `outer` and `res` are indeed the only **required** parameters. 

Every new user's **first time** using this function will be on simple polygon with no holes. If they had to guess, they'd probably try something like `polyfill_polygon(latlngs, res)`. Option 4 is nice in that this first try (and probably the most common future use-case) *just works*.

A new user's **second time** using the function might involve a hole or two, but at that point, we can expect them to read the docstring ;)

Reply:
if I can jump in with another option:

`polyfill_polygon(coords, res, order='latlng')` with an internal test for the array depth so you can pass in either `outer` or `[outer, hole1, hole2, ...]`

Reply:
Eg:

```py
def polyfill_polygon(coords, res, order='latlng'):
  if isinstance(coords[0][0], (int, float)): coords = [coords]
  ...
```

Reply:
One vote for AJs #4 `polyfill_polygon(outer, res, holes=None, order='latlng')`

Reply:
FWIW, the JS bindings follow @dfellis's suggestion - we allow either a single polygon loop, or a nested polygon loop, and check the input depth to determine which was provided.

Reply:
@dfellis That's another good option. I think the downsides are:

- it adds hacky type checking code (e.g., do we only check the first element?), and
- it makes it harder to understand what's going on when reading code that's calling the function; i think option 4 [makes it more explicit](https://www.python.org/dev/peps/pep-0020/).

Reply:
Also strong support for `res=None`, rather than providing a default. This is an easy param to forget, and I'd rather see a hard failure if it's left out.

Reply:
> @dfellis That's another good option. I think the downsides are:
> 
>     * it adds hacky type checking code (e.g., do we only check the first element?), and
> 
>     * it makes it harder to understand what's going on when reading code that's calling the function; i think option 4 [makes it more explicit](https://www.python.org/dev/peps/pep-0020/).

On your first point, I don't think checking the first element is any worse than the current behavior of not checking anything at all?

On your second point, I think that's more a style issue than anything. You could consider the first parameter to be the polygon they want to render, where the type of the polygon can basically be `Number[][] | Number[][][]` depending on if there are holes or not, while your approach requires the end user to do their own introspection into the polygon object they're working with and split it into two variables.

Reply:
> On your first point, I don't think checking the first element is any worse than the current behavior of not checking anything at all?

The logic would exist to disambiguate the input format. If we do option 4, then there's no need for that code, because we only allow one format.

> On your second point, I think that's more a style issue than anything. You could consider the first parameter to be the polygon they want to render, where the type of the polygon can basically be `Number[][] | Number[][][]` depending on if there are holes or not, while your approach requires the end user to do their own introspection into the polygon object they're working with and split it into two variables.

True, but I kind of like the idea of expecting the user to know what input format they're using. I also think the function signature of option 4 is more self-documenting than if we allowed a general `coords` input.

I also think option 4 is a more natural format for the data. The `Number[][][]` type is more of a weird artifact of the GeoJSON spec. Option 4 better matches the underlying C struct.

Whatever we do, we'll always need a code path like option 4, even if it is just a helper function that we run after disambiguating the input.

I'll try to rework things in the form of 4. We can revisit before merging (and maybe think of better function names...)